src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,382 @@
+    fn close_project(&self, project_id:Uuid) -> ();
+
+    /// Requests that the project picker lists the user's most recently opened
+    /// projects.
+    #[MethodInput=ListRecentInput,rpc_name="project/listRecent",result=list_recent_result,
+    set_result=set_list_recent_result]
+    fn list_recent(&self, number_of_projects:u32) -> ProjectListResponse;
+
+    /// Requests the creation of a new project.
+    #[MethodInput=CreateProjectInput,rpc_name="project/create",result=create_project_result,set_result=set_create_project_result]
+    fn create_project(&self, name:String) -> CreateProjectResponse;
+
+    /// Requests the deletion of a project.
+    #[MethodInput=DeleteProjectInput,rpc_name="project/delete",result=delete_project_result,set_result=set_delete_project_result]
+    fn delete_project(&self, project_id:Uuid) -> ();
+
+    /// Requests a list of sample projects that are available to the user.
+    #[MethodInput=ListSampleInput,rpc_name="project/listSample",result=list_sample_result,set_result=set_list_sample_result]
+    fn list_sample(&self, num_projects:u32) -> ProjectListResponse;
+}
+}
+
+/// IP address with host and port.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+pub struct IpWithSocket {
+    host : String,
+    port : u16
+}
+
+/// Project information, such as name, its id and last time it was opened.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+pub struct ProjectMetaData {
+    name        : String,
+    id          : Uuid,
+    last_opened : Option<UTCDateTime>
+}
+
+/// Response of `list_recent` and `list_sample`.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+pub struct ProjectListResponse {
+    /// List of projects.
+    pub projects : Vec<ProjectMetaData>
+}
+
+/// Response of `create_project`.
+#[derive(Debug,Clone,Copy,Serialize,Deserialize,PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct CreateProjectResponse {
+    /// Created project uuid.
+    pub project_id : Uuid
+}
+
+/// Response of `open_project`.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct OpenProjectResponse {
+    /// Language server address.
+    pub language_server_address : IpWithSocket
+}
+
+
+
+// ========================
+// === MockClient tests ===
+// ========================
+
+#[cfg(test)]
+mod mock_client_tests {
+    use super::API;
+    use super::MockClient;
+    use super::IpWithSocket;
+    use super::ProjectMetaData;
+    use chrono::DateTime;
+    use uuid::Uuid;
+    use json_rpc::error::RpcError;
+    use json_rpc::messages::Error;
+    use json_rpc::Result;
+    use std::future::Future;
+    use utils::test::poll_future_output;
+    use crate::project_manager::{ProjectListResponse, CreateProjectResponse, OpenProjectResponse};
+
+    fn error<T>(message:&str) -> Result<T> {
+        let err = Error {
+            code : 1,
+            data : None,
+            message : message.to_string()
+        };
```

Comment:
change it to vars and make the code shorter please.

Reply:
Again, we discussed this a long time ago, and I completely disagree. When you make vars, It will be shorter only one line, but much less readable.

Reply:
I disagree with your disagreement! But yeah, ok, lets leave it as is.

Reply:
I agree with the agreement of @notdanilo ❤️

Reply:
Actually I don't remember that this was a standard... and at least the standard the team had agreement on. And there is no about it in our coding style.